src/scores.jl

```diff
@@ -472,11 +472,83 @@ function get_score_values(chain_data::ChainScoreData, score_name::String)
 end
 
 
+function district_and_plan_score_names(chain_data::ChainScoreData)::Array{String, 1}
+    """ Simple helper function for `save_scores`. Extracts the names of all of
+        the district/plan-level scores (including those nested within
+        CompositeScores).
+    """
+    score_names = String[]
+    for score in chain_data.scores
+        if score isa CompositeScore # add children score names
+            named_scores = filter(s -> isdefined(s, :name), score.scores)
+            append!(score_names, [s.name for s in named_scores])
+        else
+            push!(score_names, score.name)
+        end
+    end
+    return score_names
+end
+
+
 function save_scores(filename::String,
-                     chain_data::ChainScoreData)
-    """ Save the `scores` in a JSON file named `filename`.
+                     chain_data::ChainScoreData,
+                     score_names::Array{String,1}=String[])
+    """ Save the `scores` in a CSV file named `filename`. In order to avoid
+        loading all the score values into RAM (which was the point of using
+        delta scores), we instead iterate through the history of the chain
+        and write the CSV row-by-row.
     """
     open(filename, "w") do f
-        JSON.print(f, chain_data.step_values)
+        if isempty(score_names) # by default, export all scores from chain
+            score_names = district_and_plan_score_names(chain_data)
+        end
+
+        column_names = String[] # colum names of the CSV
+        nested_keys = Dict{String, String}() # map score key to nested key, if any
+
+        num_districts = nothing
+        for score_name in score_names
+            score, nested_key = get_score_by_name(chain_data, score_name)
+            if isnothing(score)
+                throw(KeyError("No score in the ChainScoreData object matches the name: $score_name"))
+            elseif score isa DistrictScore || score isa DistrictAggregate
+                if isnothing(num_districts)
+                    initial_vals = chain_data.step_values[1]
+                    if !isnothing(nested_key)
+                        initial_vals = initial_vals[nested_key]
+                    end
+                    num_districts = length(initial_vals[score_name])
+                end
+                # add new column for each district
+                district_columns = ["$(score_name)_$(i)" for i in 1:num_districts]
```

Comment:
Hmm... I suppose this is as good as it gets with CSVs—the only alternative I can think of is storing array literals with _n_ entries rather than creating _n_ separate columns, but I don't think that's conventional. These files are going to be annoying to look at for a sufficiently large number of districts. Ugh.

Reply:
toootally agree with you on that. but i couldn't think of a better way!